watch-library/hardware/watch/watch_private_cdc.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Joey Castillo
+ * Copyright (c) 2023 Edward Shin
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "watch_private_cdc.h"
+
+#include "watch_utility.h"
+
+#include "tusb.h"
+
+/*
+ * Basic write buffer for USB CDC serial.
```

Comment:
What would happen if we didn't have this buffer and just relied on TinyUSB's buffer? i.e. just did tud_cdc_write immediately in _write.

If the reason is that there's some kind of relationship because tud_task and cdc_task, then it might be a problem that their calling locations have now been separated.

Reply:
That was my initial approach, but I just wasn't able to get it working reliably. I suspect it should be possible, but `tud_cdc_write()` does have the side effect of potentially dropping into USB transfer code when it flushes, which I suspect was triggering some programmatic errors.

As for `tud_task`, and `cdc_task`, they're intentionally separated, as it turns out the tud user-facing API should not be called from an interrupt context. I originally had it all serviced in the TC0 ISR, and it started behaving much better once I broke it out.